lib/kubernetes/client.rb

```diff
@@ -0,0 +1,123 @@
+# frozen_string_literal: true
+
+class Kubernetes::Client
+  SERVICE_FINALIZER = "k8s.ubicloud.com/load-balancer"
+
+  def initialize(kubernetes_cluster, session)
+    @session = session
+    @kubernetes_cluster = kubernetes_cluster
+  end
+
+  def lb_name_for_svc
+    @kubernetes_cluster.ubid
+  end
+
+  def is_service_deleted(svc)
+    svc.dig("metadata", "deletionTimestamp")
+  end
+
+  def is_service_finalized(svc)
+    svc.dig("metadata", "finalizers")&.include?(SERVICE_FINALIZER)
+  end
+
+  # Returning an array in this format:
+  # [[src_port_0, dst_port_0], [src_port_1, dst_port_1],...]
+  def lb_desired_ports(svc)
+    svc["spec"]["ports"].map { |port| [port["port"], port["nodePort"]] }
+  end
+
+  def is_load_balancer_hostname_set(svc)
+    !svc.dig("status", "loadBalancer", "ingress", "hostname").to_s.empty?
+  end
+
+  def kubectl(cmd)
+    @session.exec!("sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf #{cmd}")
+  end
+
+  def set_load_balancer_host_name(svc, hostname)
+    kubectl("-n #{svc["metadata"]["namespace"]} patch service #{svc["metadata"]["name"]} --type='merge' -p '{\"status\":{\"loadBalancer\":{\"ingress\":[{\"hostname\":\"#{hostname}\"}]}}}' --subresource='status'")
+  end
+
+  def add_svc_finalizer(svc)
+    kubectl("-n #{svc["metadata"]["namespace"]} patch service #{svc["metadata"]["name"]} --type='merge' -p '{\"metadata\":{\"finalizers\":[\"#{SERVICE_FINALIZER}\"]}}'")
+  end
+
+  def remove_svc_finalizer(svc)
+    kubectl("-n #{svc["metadata"]["namespace"]} patch service #{svc["metadata"]["name"]} --type='json' -p='[{\"op\": \"remove\", \"path\": \"/metadata/finalizers\",\"value\":[\"#{SERVICE_FINALIZER}\"]}]'")
```

Comment:
This is a bit "ew", rather than a JSON.generate or a , unless you feel it's much better to have it like this (e.g. to cut-and-paste into the terminal). Seems like it's asking for an ungratifying (although quick to fix) typo.